detection/opencv_detector.go

```diff
@@ -0,0 +1,396 @@
+//go:build opencv
+// +build opencv
+
+package detection
+
+import (
+	"fmt"
+	"image"
+	"log"
+	"math"
+	"os"
+	"path/filepath"
+	"sort"
+	"strings"
+
+	"gocv.io/x/gocv"
+)
+
+// OpenCVDetector provides native Go OpenCV object detection capabilities
+type OpenCVDetector struct {
+	net         *gocv.Net
+	classes     []string
+	initialized bool
+}
+
+// DetectionBox represents a detected object with its bounding box and confidence
+type DetectionBox struct {
+	ClassID    int
+	Class      string
+	Confidence float32
+	X          int
+	Y          int
+	Width      int
+	Height     int
+}
+
+// NewOpenCVDetector creates a new OpenCV detector instance
+func NewOpenCVDetector() *OpenCVDetector {
+	detector := &OpenCVDetector{}
+	detector.initialize()
+	return detector
+}
+
+// initialize sets up the YOLO detector with pre-trained weights
+func (d *OpenCVDetector) initialize() {
+	// Try to load YOLO model files in order of preference
+	modelPaths := []struct {
+		weights string
+		config  string
+		name    string
+	}{
+		{"/opt/yolo/yolov3-tiny.weights", "/opt/yolo/yolov3-tiny.cfg", "YOLOv3-tiny"},
+		{"/opt/yolo/yolov3.weights", "/opt/yolo/yolov3.cfg", "YOLOv3"},
+		{"/opt/yolo/yolov4-tiny.weights", "/opt/yolo/yolov4-tiny.cfg", "YOLOv4-tiny"},
+	}
+
+	classesPath := "/opt/yolo/coco.names"
+
+	// Load class names
+	if err := d.loadClassNames(classesPath); err != nil {
+		log.Printf("Failed to load class names from %s: %v", classesPath, err)
+		// Use default COCO class names as fallback
+		d.classes = cocoClassNames
+	}
+
+	// Try each model configuration
+	for _, model := range modelPaths {
+		if d.tryLoadModel(model.weights, model.config, model.name) {
+			d.initialized = true
+			return
+		}
+	}
+
+	log.Printf("Warning: No YOLO models found. OpenCV detection will not be available.")
+	d.initialized = false
+}
+
+// tryLoadModel attempts to load a specific YOLO model
+func (d *OpenCVDetector) tryLoadModel(weightsPath, configPath, modelName string) bool {
+	// Check if both files exist
+	if _, err := os.Stat(weightsPath); os.IsNotExist(err) {
+		return false
+	}
+	if _, err := os.Stat(configPath); os.IsNotExist(err) {
+		return false
+	}
+
+	// Try to load the network
+	net := gocv.ReadNetFromDarknet(configPath, weightsPath)
+	if net.Empty() {
+		log.Printf("Failed to load %s model from %s", modelName, weightsPath)
+		return false
+	}
```

Comment:
how much RAM will be needed to load the network? will this fit into the Pi's main memory or is it expected to involve swapping from RAM to disk?